spr.py

```diff
@@ -0,0 +1,250 @@
+import srp
+import os
+import time
+
+# As per PySRP sample: Consider enabling RFC5054 compatibility for interoperation with non pysrp SRP-6a implementations
+srp.rfc5054_enable()
+
+
+class AuthenticationFailed(Exception):
+    pass
+
+
+# UserID is presumably gigya_person_id that can be found if execute "renault-api settings". The same ID is sent at the srp-set endpoint by android app, so probably that the right one
+UserID = "aaaaa11-2222-3333-4444-abcabcabc123"
+# AccountId is what returned if execute "renault-api accounts"
+AccountId = "eeeaa11-2222-3333-4444-abcabcabc000"
+# PinCode is the password used in SRP, and the only password that is requested in android in this flow is the Pin code.
+PinCode = "0000"
+# vin of the car
+vin = "X7LHABC1234567890"
+# country code
+Country = "US"
+
+usr = srp.User(UserID, PinCode)
+
+uname, A = usr.start_authentication()
+ParsedID = ""
+ParsedA = ""
+ParsedB = ""
+ParsedSalt = ""
+ParsedResponse = ""
+M = b""
+
+
+def res_state():
+    print("")
+    print("========================")
+    print("1. RES-STATE")
+    print("========================")
+    strExec = (
+        'renault-api --debug http get "/commerce/v1/accounts/'
+        + AccountId
+        + "/kamereon/kca/car-adapter/v1/cars/"
+        + vin
+        + '/res-state?country='
+        + Country
+        + '"'
+    )
+    stream = os.popen(strExec)
+
+
+def srp_set():
+    ParseList1 = []
+    ParseList2 = []
+    global ParsedID
+    print("")
+    print("========================")
+    print("2. SRP-SET ")
+    print("========================")
+
+    strExec = (
+        "renault-api --debug http post '/commerce/v1/accounts/"
+        + AccountId
+        + "/kamereon/kca/car-adapter/v1/cars/"
+        + vin
+        + '/actions/srp-sets?country=' + Country + '\' \'{"data":{"type":"SrpSets","attributes":{"a":"'
+        + A.hex().upper()
+        + '","i":"'
+        + UserID
+        + "\"}}}'"
+    )
+
+    stream = os.popen(strExec)
+    # output = stream.read()
+    time.sleep(5)
+
+    output = " ".join(stream)
+    output = output.replace('"', "")
+    output = output.replace("'", "")
+    output = output.replace(" ", "")
+
+    ParseList1 = output.split("SrpSets,id:")
```

Comment:
I think this is very difficult to understand.
You should use JSON parsing utilities instead to convert it into an object.

Reply:
Yeah I know - I am just comparing your script with the traces you sent me previously and your parsing/splitting makes it hard to reconcile.
Maybe you need to adjust your commands to use `renault-api --log --json http get/post`
This should create log files (easier to share) and return json (easier to parse)

Reply:
Ok, so then I can just describe what is being parsed then:
1. First step is GET res-state endpoint. Again not sure what it does, maybe resets the flow, maybe checks that engine is not started already. We don't parse there anything.
2. At this step we start SRP auth. We generate so called "A" for srp based on SRP username and password, and send "A" in POST srp-sets. The length of this thing is 512 and servers checks this fact. For the POST to srp-sets, the server replies with the Notification ID, where we will then find the server result. So we need to parse the "id" in the bodyof POST response.
3. Now we GET kmr notification endpoint with the id we received above. If the body response has Salt and LoginB, we try to parse them, if not - try the same GET kmr request again after a delay. Those Salt and LoginB are the Server's response for SRP, these are used to calculate the final SRP key (M in the code, but also called "srp" in the following http request)
4. When we calculated M we can send it in the POST to engine-start. The request body has the command and M value inside the "srp" field. This is where the server can completely verify srp auth. Just like at the step 2, we don't get the server response immediately, but instead we get the "id" of the notification to check.
5. We request the notification with that new id and see the response if it fails or not.

Reply:
I was able to work out all that, but it is still much harder to read than if you used json :)

Reply:
I mean, come on! I basicaly googled every other line to write in python, don't make me do any of that :)